File: apps/neurolinguist-json-parser/api/bigquery-export.py
"""BigQuery Export API Endpoint

This serverless function handles direct export of validated entities to BigQuery.
It follows the DATA PROTECTION LAWS established in the truth_forge project.

Per DATA PROTECTION LAWS:
- Uses batch load (never streaming)
- Creates backup before write
- Validates schema alignment
- Reports all errors

Author: Claude (Opus 4.5)
Date: 2026-02-02
"""
from __future__ import annotations

import json
import logging
import os
import tempfile
from datetime import UTC, datetime
from http.server import BaseHTTPRequestHandler
from typing import Any

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Target table
TABLE_ID = "flash-clover-464719-g1.spine.entity_unified"


def validate_entity(entity: dict[str, Any], index: int) -> list[str]:
    """Validate a single entity against the schema.

    Returns list of error messages (empty if valid).
    """
    errors = []

    # Required fields
    required_fields = [
        "entity_id", "level", "entity_type", "entity_mode",
        "conversation_id", "text", "source_pipeline", "source_system",
        "content_date", "created_at", "updated_at", "ingestion_timestamp",
        "ingestion_job_id", "validation_status"
    ]

    for field in required_fields:
        if field not in entity or entity[field] is None:
            errors.append(f"Entity {index}: Missing required field '{field}'")

    # Validate level
    if "level" in entity:
        level = entity["level"]
        if level not in [2, 3, 4, 5, 6, 7, 8]:
            errors.append(f"Entity {index}: Invalid level {level} (must be 2-8)")

    # Validate entity_id format
    if "entity_id" in entity:
        entity_id = entity["entity_id"]
        if not isinstance(entity_id, str) or ":" not in entity_id:
            errors.append(f"Entity {index}: Invalid entity_id format '{entity_id}'")

    # Validate validation_status
    if "validation_status" in entity:
        status = entity["validation_status"]
        if status not in ["PASSED", "WARNING", "FAILED"]:
            errors.append(f"Entity {index}: Invalid validation_status '{status}'")

    # Validate parent_id for non-L8 entities
    if "level" in entity and entity["level"] < 8:
        if "parent_id" not in entity or entity["parent_id"] is None:
            errors.append(f"Entity {index}: Non-L8 entity missing parent_id")

    return errors


def validate_batch(entities: list[dict[str, Any]]) -> tuple[bool, list[str]]:
    """Validate a batch of entities.

    Returns (is_valid, error_messages).
    """
    all_errors = []

    for i, entity in enumerate(entities):
        errors = validate_entity(entity, i)
        all_errors.extend(errors)

    # Check for duplicate entity_ids
    entity_ids = [e.get("entity_id") for e in entities if e.get("entity_id")]
    duplicates = set([x for x in entity_ids if entity_ids.count(x) > 1])
    if duplicates:
        all_errors.append(f"Duplicate entity_ids found: {list(duplicates)[:5]}...")

    return len(all_errors) == 0, all_errors


def write_to_bigquery(entities: list[dict[str, Any]]) -> dict[str, Any]:
    """Write entities to BigQuery using batch load.

    Returns result dict with success status and details.
    """
    try:
        from google.cloud import bigquery
        from google.cloud.bigquery import (
            LoadJobConfig,
            SourceFormat,
            WriteDisposition,
            CreateDisposition,
        )
    except ImportError:
        return {
            "success": False,
            "error": "google-cloud-bigquery not installed",
            "rows_written": 0,
        }

    try:
        client = bigquery.Client()

        # Create backup first (DATA PROTECTION LAWS)
        timestamp = datetime.now(UTC).strftime("%Y%m%d_%H%M%S")
        backup_table_id = f"{TABLE_ID}_backup_{timestamp}"

        backup_query = f"""
        CREATE TABLE `{backup_table_id}`
        OPTIONS (expiration_timestamp = TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 7 DAY))
        AS SELECT * FROM `{TABLE_ID}`
        """

        logger.info(f"Creating backup: {backup_table_id}")
        backup_job = client.query(backup_query)
        backup_job.result()

        # Verify backup
        backup_table = client.get_table(backup_table_id)
        backup_rows = backup_table.num_rows
        logger.info(f"Backup created with {backup_rows} rows")

        # Write entities to temp file
        with tempfile.NamedTemporaryFile(
            mode="w",
            suffix=".ndjson",
            delete=False,
            prefix="bq_export_",
        ) as f:
            for entity in entities:
                # Convert metadata dict to JSON string
                entity_copy = entity.copy()
                if isinstance(entity_copy.get("metadata"), dict):
                    entity_copy["metadata"] = json.dumps(entity_copy["metadata"])
                f.write(json.dumps(entity_copy) + "\n")
            temp_path = f.name

        # Configure batch load job
        job_config = LoadJobConfig(
            source_format=SourceFormat.NEWLINE_DELIMITED_JSON,
            write_disposition=WriteDisposition.WRITE_APPEND,
            create_disposition=CreateDisposition.CREATE_NEVER,
        )

        # Execute batch load
        with open(temp_path, "rb") as source_file:
            load_job = client.load_table_from_file(
                source_file,
                TABLE_ID,
                job_config=job_config,
            )

        load_job.result()

        if load_job.errors:
            return {
                "success": False,
                "error": f"Load job errors: {load_job.errors}",
                "rows_written": 0,
                "backup_table": backup_table_id,
            }

        rows_written = load_job.output_rows or len(entities)

        # Cleanup temp file
        try:
            os.unlink(temp_path)
        except Exception:
            pass

        return {
            "success": True,
            "rows_written": rows_written,
            "backup_table": backup_table_id,
            "backup_rows": backup_rows,
            "table": TABLE_ID,
        }

    except Exception as e:
        logger.error(f"BigQuery write failed: {e}")
        return {
            "success": False,
            "error": str(e),
            "rows_written": 0,
        }


class handler(BaseHTTPRequestHandler):
    """Vercel serverless function handler."""

    def do_OPTIONS(self):
        """Handle CORS preflight."""
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def do_POST(self):
        """Handle POST request to export entities."""
        # CORS headers
        self.send_header("Access-Control-Allow-Origin", "*")

        try:
            # Read request body
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            data = json.loads(body.decode("utf-8"))

            entities = data.get("entities", [])
            dry_run = data.get("dry_run", False)

            if not entities:
                self.send_response(400)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({
                    "success": False,
                    "error": "No entities provided",
                }).encode())
                return

            logger.info(f"Received {len(entities)} entities (dry_run={dry_run})")

            # Validate batch
            is_valid, errors = validate_batch(entities)

            if not is_valid:
                self.send_response(400)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({
                    "success": False,
                    "error": "Validation failed",
                    "validation_errors": errors[:20],  # Limit error count
                    "total_errors": len(errors),
                }).encode())
                return

            # If dry run, just return validation success
            if dry_run:
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({
                    "success": True,
                    "dry_run": True,
                    "validated_entities": len(entities),
                    "message": "Validation passed. Ready for export.",
                }).encode())
                return

            # Write to BigQuery
            result = write_to_bigquery(entities)

            if result["success"]:
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps(result).encode())
            else:
                self.send_response(500)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps(result).encode())

        except json.JSONDecodeError as e:
            self.send_response(400)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({
                "success": False,
                "error": f"Invalid JSON: {e}",
            }).encode())

        except Exception as e:
            logger.error(f"Handler error: {e}")
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({
                "success": False,
                "error": f"Server error: {e}",
            }).encode())

File: apps/neurolinguist-json-parser/services/bigqueryExport.ts
/**
 * BigQuery Export Service
 * Handles direct export of validated entities to BigQuery via serverless API
 *
 * This service follows the DATA PROTECTION LAWS:
 * - Validates entities before export
 * - Uses batch load (never streaming)
 * - Creates backup before write
 * - Reports all errors
 */

import type { Conversation } from '../types';
import { exportToNativeSpineFormat, NativeSpineEntity } from './spineExport';
import { validateConversation, ValidationResult } from './exportValidation';

/**
 * BigQuery export result
 */
export interface BigQueryExportResult {
  success: boolean;
  error?: string;
  rows_written?: number;
  backup_table?: string;
  backup_rows?: number;
  table?: string;
  validation_errors?: string[];
  total_errors?: number;
}

/**
 * Export progress callback
 */
export type ExportProgressCallback = (progress: {
  stage: 'validating' | 'preparing' | 'uploading' | 'complete' | 'error';
  message: string;
  percentage: number;
}) => void;

/**
 * API base URL - uses relative path for same-origin, or configurable for dev
 */
const API_BASE_URL = import.meta.env.VITE_API_URL || '/api';

/**
 * Validate entities for BigQuery export (dry run)
 */
export async function validateForBigQuery(
  conversation: Conversation,
  options: {
    sourceFile?: string;
    includeWords?: boolean;
  } = {}
): Promise<{
  valid: boolean;
  entityCount: number;
  validationResult: ValidationResult;
  errors: string[];
}> {
  // First, validate conversation structure
  const validationResult = validateConversation(conversation);

  if (!validationResult.valid) {
    return {
      valid: false,
      entityCount: 0,
      validationResult,
      errors: validationResult.issues.map(i => `${i.code}: ${i.message}`),
    };
  }

  // Generate entities to validate
  const exportResult = await exportToNativeSpineFormat(conversation, {
    sourceFile: options.sourceFile,
    includeWords: options.includeWords ?? true,
  });

  // Call API in dry_run mode
  try {
    const response = await fetch(`${API_BASE_URL}/bigquery-export`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        entities: exportResult.entities,
        dry_run: true,
      }),
    });

    const result = await response.json();

    return {
      valid: result.success,
      entityCount: exportResult.entities.length,
      validationResult,
      errors: result.validation_errors || [],
    };
  } catch (error) {
    return {
      valid: false,
      entityCount: exportResult.entities.length,
      validationResult,
      errors: [`API error: ${error instanceof Error ? error.message : 'Unknown error'}`],
    };
  }
}

/**
 * Export conversation directly to BigQuery
 */
export async function exportToBigQuery(
  conversation: Conversation,
  options: {
    sourceFile?: string;
    includeWords?: boolean;
    onProgress?: ExportProgressCallback;
  } = {}
): Promise<BigQueryExportResult> {
  const { onProgress } = options;

  try {
    // Stage 1: Validate
    onProgress?.({
      stage: 'validating',
      message: 'Validating conversation structure...',
      percentage: 10,
    });

    const validationResult = validateConversation(conversation);
    if (!validationResult.valid) {
      const errors = validationResult.issues
        .filter(i => i.level === 'error')
        .map(i => i.message);

      return {
        success: false,
        error: `Validation failed: ${validationResult.stats.errors} errors`,
        validation_errors: errors,
        total_errors: validationResult.stats.errors,
      };
    }

    // Stage 2: Prepare entities
    onProgress?.({
      stage: 'preparing',
      message: 'Generating BigQuery-compatible entities...',
      percentage: 30,
    });

    const exportResult = await exportToNativeSpineFormat(conversation, {
      sourceFile: options.sourceFile,
      includeWords: options.includeWords ?? true,
    });

    const entityCount = exportResult.entities.length;

    // Stage 3: Upload to BigQuery
    onProgress?.({
      stage: 'uploading',
      message: `Uploading ${entityCount.toLocaleString()} entities to BigQuery...`,
      percentage: 50,
    });

    const response = await fetch(`${API_BASE_URL}/bigquery-export`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        entities: exportResult.entities,
        dry_run: false,
      }),
    });

    const result: BigQueryExportResult = await response.json();

    if (result.success) {
      onProgress?.({
        stage: 'complete',
        message: `Successfully exported ${result.rows_written?.toLocaleString()} rows`,
        percentage: 100,
      });
    } else {
      onProgress?.({
        stage: 'error',
        message: result.error || 'Export failed',
        percentage: 100,
      });
    }

    return result;
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error';

    onProgress?.({
      stage: 'error',
      message: `Export failed: ${errorMessage}`,
      percentage: 100,
    });

    return {
      success: false,
      error: errorMessage,
    };
  }
}

/**
 * Check if BigQuery export API is available
 */
export async function checkBigQueryApiStatus(): Promise<{
  available: boolean;
  message: string;
}> {
  try {
    const response = await fetch(`${API_BASE_URL}/bigquery-export`, {
      method: 'OPTIONS',
    });

    if (response.ok) {
      return {
        available: true,
        message: 'BigQuery export API is available',
      };
    }

    return {
      available: false,
      message: `API returned status ${response.status}`,
    };
  } catch (error) {
    return {
      available: false,
      message: `API not reachable: ${error instanceof Error ? error.message : 'Unknown error'}`,
    };
  }
}

/**
 * Get export statistics for a conversation
 */
export async function getExportStats(
  conversation: Conversation,
  options: { includeWords?: boolean } = {}
): Promise<{
  totalEntities: number;
  byLevel: Record<number, number>;
  estimatedSize: string;
}> {
  const exportResult = await exportToNativeSpineFormat(conversation, {
    includeWords: options.includeWords ?? true,
  });

  const byLevel: Record<number, number> = {};
  for (const entity of exportResult.entities) {
    byLevel[entity.level] = (byLevel[entity.level] || 0) + 1;
  }

  // Estimate size (rough: ~500 bytes per entity average)
  const estimatedBytes = exportResult.entities.length * 500;
  let estimatedSize: string;

  if (estimatedBytes < 1024) {
    estimatedSize = `${estimatedBytes} B`;
  } else if (estimatedBytes < 1024 * 1024) {
    estimatedSize = `${(estimatedBytes / 1024).toFixed(1)} KB`;
  } else {
    estimatedSize = `${(estimatedBytes / (1024 * 1024)).toFixed(1)} MB`;
  }

  return {
    totalEntities: exportResult.entities.length,
    byLevel,
    estimatedSize,
  };
}
